Validate badge selection before input is saved.

matjaeck opened the pull request:
Fixes OpenRA/OpenRAWeb#396.

matjaeck commented:
> pchote> lawANDorder: i have a strong feeling that most of your validation code could go away.

Submitting the selection works in 3 steps:

1. Delete all badges from the user badge table that are not in the array of marked badges or delete all if none is marked.

1. For each marked badge, check if there is already one with the same id in the user badge table.

1. Insert the new badge in the user table if it is not a duplicate.

I don't want to delete anything before the selection is validated so the current code setup requires to validate (=check if it's a default badge, if not, check if the user has the required availabilities) the selection before the insert loop.

matjaeck commented:
Added an error when somebody tries to manipulate the form to revoke keys of other users.

matjaeck commented:
Updated again. Now using the the build_query function and type casting for integers.

matjaeck commented:
Updated.

pchote commented:
Can you instead do something like this to simplify the logic and reduce the query count?
*Note: I haven't tested this at all, so there may be syntax or logic errors.*

```php
if ($submitmarked)
{
	// Check form key.
	if (!check_form_key($form_key))
	{
		trigger_error($this->language->lang('FORM_INVALID') . '<br><br>' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
	}

	$delete_sql = empty($markedbadges) ? '' : ' AND ' . $this->db->sql_in_set('badge_id', $markedbadges, true); // true for NOT IN, false (default) for IN

	// Delete all user badges that are NOT IN the array of marked badges or all if none is selected.
	$sql = 'DELETE FROM ' . $user_badge_table . '
		WHERE user_id = ' . $this_user_id . "
		$delete_sql";
	$this->db->sql_query($sql);

	// Query valid badges
	$sql = "SELECT badge_id
		FROM $badge_table
		WHERE badge_default = TRUE";
	if (!($result = $this->db->sql_query($sql)))
	{
		trigger_error($this->user->lang('UCP_AVAIL_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
	}
	$default_badges = $this->db->sql_fetchrowset($result);

	$sql = "SELECT badge_id
		FROM $badge_avail_table
		WHERE user_id = $this_user_id";						
	if (!($result = $this->db->sql_query($sql)))
	{
		trigger_error($this->user->lang('UCP_AVAIL_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
	}
	$custom_badges = $this->db->sql_fetchrowset($result);

	$sql_ary = array();
	foreach ($markedbadges as $markedbadge)
	{
		// Validate requested badges
		if (!in_array($markedbadge, $default_badges) && !in_array($markedbadge, $custom_badges))
			continue;

		$sql_ary[] = array(
			'user_id' => $this_user_id,
			'badge_id' => $markedbadge,
			'badge_order' => null // Order value will be validated after loop.
		);
	}

	if (!$db->sql_multi_insert($user_badge_table, $sql_ary))
	{
		trigger_error($this->user->lang('UCP_DUPLICATE_USER_BADGE_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
	}

	if (!$this->core->validate_badge_order($this_user_id))
	{
		trigger_error($this->user->lang('UCP_BADGE_ORDER_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
	}
	unset($markedbadges);
	meta_refresh(5, $this->u_action);
	trigger_error($this->user->lang('UCP_SELECTED_BADGES_SAVED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
}
```

matjaeck commented:
Thank you for your help, I have updated the commit with the changes you suggested, however I had to change some things too avoid duplicates and wrong error messages.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -319,57 +323,97 @@ function main($id, $mode)
-							WHERE badge_id = $markedbadge
-							AND user_id = $this_user_id";
+						// Query valid badges
+						$sql = "SELECT badge_id
```
On closer read, we can squash all three queries (including the `$user_badges` above, which is only used here) into a single query that lists all and only the valid badges:
```php
// Query valid badges
$sql_ary = array(
	'SELECT'	=> 'b.badge_id',

	'FROM'		=> array(
		$badge_table		=> 'b',
	),

	'LEFT_JOIN'	=> array(
		array(
			'FROM'	=> array($user_badge_table => 'bu'),
			'ON'	=> 'b.badge_id = bu.badge_id AND bu.user_id = ' . $this_user_id
		),
		array(
			'FROM'	=> array($badge_avail_table => 'ba'),
			'ON'	=> 'b.badge_id = ba.badge_id AND ba.user_id = ' . $this_user_id
		),
	),

	// Badges are valid if they are not already selected (bu.badge_id) and are either:
	//  - a default badge (b.badge_default)
	//  - have been awarded to the user (ba.user_id)
	'WHERE'		=> 'bu.badge_id IS NULL AND (b.badge_default = TRUE OR ba.user_id = ' . $this_user_id . ')',
);

$sql = $this->db->sql_build_query('SELECT', $sql_ary);
```

The check below then simplifies to `if ($this->core->in_2d_array($markedbadge, $valid_badges))`

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -319,57 +323,97 @@ function main($id, $mode)
+							trigger_error($this->user->lang('UCP_DEFAULT_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
 						}
-						$duplicates = (int) $this->db->sql_fetchfield('duplcount');
+						$default_badges = $this->db->sql_fetchrowset($result);
```
On closer inspection it would be better to replace sql_fetchrowset with:
```php
$valid_badges = array();
while ($row = $this->db->sql_fetchrow($result))
{
	$valid_badges[] = $row['badge_id'];
}
```

The check below then simplifies even further to `if (in_array($markedbadge, $valid_badges))`

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -319,57 +323,97 @@ function main($id, $mode)
+							}
+						}
+
+						if(!empty($sql_array))
```
Style nit: need a space between `if (`

pchote replied:
On closer inspection `$this->db->sql_multi_insert` already includes a check for empty input, so this whole block can simplify down to `$this->db->sql_multi_insert($user_badge_table, $sql_array);` if we also drop the failure check (which upstream phpbb code also doesn't do).
```php

```

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -319,57 +323,97 @@ function main($id, $mode)
+
+						if (!$this->core->validate_badge_order($this_user_id))
+						{
+							trigger_error($this->user->lang('UCP_BADGE_ORDER_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
```
Do we really need to check this?  `validate_badge_order` will only return false if the initial badge query fails, but I think phpBB will generate its own error page on most of these cases?

pchote commented:
The queries for building the selection table can similarly simplify down:
```php
// Query badges for the table view
$sql_ary = array(
	'SELECT'	=> 'b.badge_id, b.badge_label, b.badge_icon_24, bt.badge_type_name, bu.badge_id IS NOT NULL as badge_selected',

	'FROM'		=> array(
		$badge_table		=> 'b',
		$badge_type_table 	=> 'bt'
	),

	'LEFT_JOIN'	=> array(
		array(
			'FROM'	=> array($user_badge_table => 'bu'),
			'ON'	=> 'b.badge_id = bu.badge_id AND bu.user_id = ' . $this_user_id
		),
		array(
			'FROM'	=> array($badge_avail_table => 'ba'),
			'ON'	=> 'b.badge_id = ba.badge_id AND ba.user_id = ' . $this_user_id
		),
	),

	// Badges are available if either:
	//  - a default badge (b.badge_default)
	//  - have been awarded to the user (ba.user_id)
	'WHERE'		=> 'b.badge_type_id = bt.badge_type_id AND (b.badge_default = TRUE OR ba.user_id = ' . $this_user_id . ')',
	'ORDER_BY'	=> 'bt.badge_type_name'
);

$sql = $this->db->sql_build_query('SELECT', $sql_ary);

if (!($result = $this->db->sql_query($sql)))
{
	trigger_error($this->user->lang('UCP_AVAIL_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
}

// Prepare an array to group badges by types.
$typelist = array();
while ($row = $this->db->sql_fetchrow($result))
{
	$s_badgerow = array(
		'S_SELECTED' => $row['badge_selected'],
		'BADGE_ICON_URL' => $row['badge_icon_24'],
		'BADGE_LABEL' => $row['badge_label'],
		'ID' => $row['badge_id'],
	);
	// Check if the type name is in the typelist, if not, add it to the list and to the blockvars for the current row.
	if (!in_array($row['badge_type_name'], $typelist))
	{
		$typelist[] = $row['badge_type_name'];
		$s_badgerow['BADGE_TYPE'] = $row['badge_type_name'];
	}
	$this->template->assign_block_vars('s_badgerow', $s_badgerow);
}

$this->db->sql_freeresult($result);

// Assign template vars for mode
$this->template->assign_vars(array(
	'S_UCP_MODE_SELECT_BADGES' => true,
	'S_AVAIL_BADGES' => sizeof($typelist) > 0
));
```

matjaeck commented:
This looks really good and will help in a lot of other places too. It's worth doing this now and I'll update the pr quickly.

matjaeck commented:
Updated.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,8 +258,8 @@ function main($id, $mode)
 				// Set up vars for mode.
-				$markedbadges = $this->request->variable('marked_badges', array(0)); // badge_ids
-				$submitmarked = $this->request->is_set_post('marked_badges') || empty($markedbadges) && $action == 'submit_marked' ? true : false;
+				$markedbadges = (array)$this->request->variable('marked_badges', array(0)); // badge_ids
```
`array(0)` is an array containing the number `0`.  Shouldn't this be an empty array (`array()`)?

matjaeck replied:
[array(0) means array has 0 index (empty)](https://stackoverflow.com/questions/13232737/var-dump-array0-php#comment18025524_13232737), with `array()` ti does not collect the values from `name="markedkeys[]" value=<int>`` in the template. phpBB uses this wherever it requests values from marked checkboxes and in a lot of other places too.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -203,10 +164,10 @@ function main($id, $mode)
-				$revokemark = $this->request->is_set_post('markedkeys') && $action == 'rev_marked' ? true : false;
-				$revokeall = $action == 'rev_all' ? true : false;
+				$start = (int)$this->request->variable('start', 0); // Used for pagination.
+				$markedkeys = (array)$this->request->variable('markedkeys', array(0)); // items ids.
```
`array(0)` is an array containing the number `0`.  Shouldn't this be an empty array (`array()`)?

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -319,143 +269,146 @@ function main($id, $mode)
 					// Delete all user badges that are NOT IN the array of marked badges or all if none is selected.
+					$delete_sql = empty($markedbadges) ? '' : ' AND ' . $this->db->sql_in_set('badge_id', $markedbadges, true); // true for NOT IN, false (default) for IN
 					$sql = 'DELETE FROM ' . $user_badge_table . '
 						WHERE user_id = ' . $this_user_id . "
```
Switching between `'`' and `"` style quotes within the same statement isn't great.  Can you please stick with `'` and concatenating in variables? This seems to be the preferred convention in most of the phpbb code.

matjaeck replied:
This is  phpBB modifies queries. Generally, `"` is more comfortable because it allows to write variables in the sql statement without any `'  . $var . '`, however `$this->user` does not work in when written in double quotes. 

I would like to keep the style for the cases when we modify queries because I'm not sure how the alternative writing would be and this is default phpBB.

pchote replied:
Why are you using `$this_user_id` with single quotes and concatenation?

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -319,143 +269,146 @@ function main($id, $mode)
-							$sql = 'INSERT INTO ' . $user_badge_table . $this->db->sql_build_array('INSERT', $data);
-							$this->db->sql_query($sql);
+							// Validate requested badges.
+							settype($markedbadge, 'int');
```
I expect this is redundant, but probably harmless.

matjaeck replied:
My idea here was that we only casted the input from `markedkeys` to an array but not its values to integers, and `$markedbadge` is used in `if (in_array($markedbadge, $valid_badges)).`

pchote replied:
Two points here:
* `in_array` will only return true if `$markedbadge` is in `$valid_badges`, which we know by definition only contains integers
* `$markedbadges` is used without explicit type checks earlier in the file (but note that `sql_in_set` escapes the entries, so this is safe)

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -464,15 +417,19 @@ function main($id, $mode)
+
+					$sql = 'SELECT badge_order as current_order
+						FROM ' . $user_badge_table . '
+						WHERE ' . $this->db->sql_build_array('SELECT', $sql_array); // No need to check for user_id, item_id is the table's primary key.
```
Using `sql_build_array` here hurts readability without any real benefit.

matjaeck replied:
`item_id` is in this case user input but it is already casted to an integer so I think it is safe to directly use the var.

matjaeck commented on ucp/useraccounts_module.php:
```diff
@@ -232,7 +193,7 @@ function main($id, $mode)
 							$sql_in = array();
```
Done.

matjaeck commented on ucp/useraccounts_module.php:
```diff
@@ -245,45 +206,34 @@ function main($id, $mode)
 
-				// Retrieve key data.
+				// Query key data.
 				$sql_array = array(
```
Done.

matjaeck commented on ucp/useraccounts_module.php:
```diff
@@ -245,45 +206,34 @@ function main($id, $mode)
-				{
-					trigger_error($this->user->lang('UCP_KEY_DATA_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
-				}
+				$result = $this->db->sql_query_limit($sql, $this->config['topics_per_page'], $start);
```
This is used by default for entries in paginated tables. We can define a custom ACP setting for it (don't really see the merit) or hardcode a value if you want.

matjaeck commented:
Updated, entirely removed the `$action` part where it is not required.

AMHOL commented:
```php
$moveup = (bool)$action == 'move_up' ? true : false;
$movedown = (bool)$action == 'move_down' ? true : false;
```

No need for casting here, value will always be a boolean, been a while since I did any PHP but pretty sure the following would suffice:

```php
$moveup = (bool)$action == 'move_up';
$movedown = (bool)$action == 'move_down';
```

Also can use the following markdown to get syntax highlighting when posting code in Github comments:

\```php
 // Code here
 \```

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -203,13 +159,14 @@ function main($id, $mode)
-				$revokeall = $action == 'rev_all' ? true : false;
+				$start = (int)$this->request->variable('start', 0); // Used for pagination.
+				$markedkeys = (array)$this->request->variable('markedkeys', array(0)); // items ids.
+				$revokemark = (bool)$this->request->is_set_post('rev_marked');
```
tbh, i'm not sure this is better than what you had before.  There can only be one action at a time, so having a single variable to define which one makes sense.  Having a separate variable for each action can open up edge cases that a malicious actor may be able to abuse.

matjaeck replied:
The $action pattern seems to appear in two usecases:
* In a switch statement where the actions are the cases (not suitable for our module with many different modes).
* When parameters are passed through the URL (we use this for ordering).
* When the confirm box is used and there is more than one possible submit "action" to confirm.

Unlike the other phpBB sections, the pattern is not as much used in the UCP context. Also, it is not mentioned in the dedicated section of [Using phpBB3's Basic Functions](https://wiki.phpbb.com/Using_phpBB3%27s_Basic_Functions#1.4.1._User_input). 

To account for your comment, I simplified all submit actions that appear in a context where only one submit actions is possible to a $submit var.

I think the current method is cleaner then the $action method which would also require to define the $action for each sumbit. Additionally, the check if the request was send via POST is done for the submit that was requested instead of checking if a random request variable in the context of the action was send via POST.

matjaeck replied:
However, I will check if the `$action` method  can be applied to match URL parameters too.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -464,15 +410,15 @@ function main($id, $mode)
+					// Get current item order value.
+					$sql = "SELECT badge_order AS current_order
 						FROM $user_badge_table
 						WHERE item_id = $item_id"; // No need to check for user_id, item_id is the table's primary key.
```
phpBB convention seems to be to cast to an int here, e.g. `WHERE item_id = ' . (int) $item_id;`

This is generally a better and safer convention because it guarantees that someone in the future can't accidentally introduce a vulnerability by removing the cast above without checking how it is used.

matjaeck replied:
Agree, however the matter has resolved itself. `request_var()` automatically casts the values of the array to integers, so it does not need to be casted here again (which was an addition by me). For this reason I removed most of the variable casting (after checking the var_dump to make sure) as they were not needed.

pchote replied:
My point above still stands, however: safe practice and as far as I can see the phpbb convention is to *always* either sql_escape or explicitly cast to an int any variables that came from user input as part of the sql query definition.

matjaeck commented:
Updated, also removed unused css definitions. Also removed 
```php
$sql_in = array();
foreach ($markedkeys as $marked)
{
	$sql_in[] = $marked;
}
```

because `$markedkeys` is already casted to an array of integers. The original code includes the code above, but IMO it is safe to skip this.

All changes are tested and when I removed casting, I ensured that the casting works by dumping the variables.

matjaeck commented:
Updated again, am now happy with the changes.

pchote commented on core/core.php:
```diff
@@ -187,4 +181,23 @@ public function validate_badge_order($user_id)
+	{
+		$sql = "SELECT COUNT(*) as count
+			FROM $table
+			WHERE user_id = $user_id";
```
Can we please cast `$user_id` to int here to guarantee safety?

pchote commented on core/core.php:
```diff
@@ -153,21 +145,23 @@ public function get_ext_img_path()
-		WHERE user_id = ' . $user_id . '
-		ORDER BY badge_order';
+			FROM ' . $this->table_prefix . 'openra_user_badges' . '
+			WHERE user_id = ' . $user_id . '
```
Can we please cast `$user_id` to int here to guarantee safety?

matjaeck replied:
I moved the query to the module, it is only used in one place anyway.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -203,39 +179,32 @@ function main($id, $mode)
 
 				// Revoke keys if requested.
-				if ($revokemark || $revokeall)
+				if ($this->request->is_set_post('markedkeys') && $revokemark || $revokeall)
```
This needs parens to denote order of operation.  I assume you mean
```php
if (($this->request->is_set_post('markedkeys') && $revokemark) || $revokeall)
```
?

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -245,45 +214,31 @@ function main($id, $mode)
+				// Pagination for the manage keys table view.
 				$sql = "SELECT COUNT(*) as keycount
 					FROM $key_table
 					WHERE user_id = $this_user_id
```
We know that `$this_user_id` is an int from above, but can we please still define this as `WHERE user_id = ' . (int)$this_user_id . '` to make it clear that this isn't a security hole?

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -245,45 +214,31 @@ function main($id, $mode)
+				// Query key data.
+				$sql = "SELECT item_id, user_id, fingerprint, registered, last_accessed, revoked
+					FROM $key_table
+					WHERE user_id = $this_user_id
```
Likewise here.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
-					// Delete all user badges that are NOT IN the array of marked badges or all if none is selected.
+					// Either delete all user badges or only those that are NOT IN the array of marked badges.
 					$sql = 'DELETE FROM ' . $user_badge_table . '
 						WHERE user_id = ' . $this_user_id . "
```
Likewise here.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
+				$markedbadges = $this->request->variable('markedbadges', array(0)); // Badge IDs, automatically casted to an array of integers.
 
-				if ($submitmarked)
+				if ($this->request->is_set_post('markedbadges') || empty($markedbadges) && $submit )
```
Need parens here too.  Mixing && and || in a statement without them is really bad practice.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
+							'LEFT_JOIN' => array(
+								array(
+									'FROM' => array($user_badge_table => 'bu'),
+									'ON' => 'b.badge_id = bu.badge_id AND bu.user_id = ' . $this_user_id
```
`(int)` here too please

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
+								),
+								array(
+									'FROM' => array($badge_avail_table => 'ba'),
+									'ON' => 'b.badge_id = ba.badge_id AND ba.user_id = ' . $this_user_id
```
And here.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
+							// Badges are valid if they are not already selected (bu.badge_id is null) and are either:
+							//  - a default badge (b.badge_default is true)
+							//  - have been awarded to the user (ba.user_id is not null)
+							'WHERE' => 'bu.badge_id IS NULL AND (b.badge_default = TRUE OR ba.user_id = ' . $this_user_id . ')'
```
And here.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
+					'LEFT_JOIN' => array(
+						array(
+							'FROM' => array($user_badge_table => 'bu'),
+							'ON' => 'b.badge_id = bu.badge_id AND bu.user_id = ' . $this_user_id
```
And here.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
+						),
+						array(
+							'FROM' => array($badge_avail_table => 'ba'),
+							'ON' => 'b.badge_id = ba.badge_id AND ba.user_id = ' . $this_user_id
```
And here.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
+					// Badges are available if either:
+					//  - a default badge (b.badge_default)
+					//  - have been awarded to the user (ba.user_id)
+					'WHERE'	=> 'b.badge_type_id = bt.badge_type_id AND (b.badge_default = TRUE OR ba.user_id = ' . $this_user_id . ')',
```
And here.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
-				$moveup = $action == 'move_up' ? true : false;
-				$movedown = $action == 'move_down' ? true : false;
+				// Set up vars for mode.
+				$action = $this->request->variable('action', ''); // Required to be defined here again, probably because parameters are given through the URL.
```
This comment gives me the heebie jeebies.  Can you please find out what is going on here, and fix it if needed?  What is the value of `$action` before and after this line?

matjaeck replied:
When using `$request->variable()`, it automatically sets a type for the variable. The code used in the else clause of 

```php
		if (is_array($action))
		{
			list($action, ) = each($action);
		}
		else
		{
			$action = $request->variable('action', '');
		}
```
is bogus and the array check IMO too so I removed both.

This makes it necessary to explicitly request the `$action` var as an array or string, like for the URL parameter:

```php
		U_MOVE_UP' => $this->u_action . '&amp;action=move_up&amp;id=' . $row['item_id']
```

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -308,154 +265,162 @@ function main($id, $mode)
-				$movedown = $action == 'move_down' ? true : false;
+				// Set up vars for mode.
+				$action = $this->request->variable('action', ''); // Required to be defined here again, probably because parameters are given through the URL.
+				$item_id = $this->request->variable('id', 0); // Item ID of selected element when reordering, automatically casted to an integer. Retrieved from URL.
```
Note: correct term is `cast` rather than `casted`.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -464,15 +429,15 @@ function main($id, $mode)
+					$sql = "SELECT badge_order AS current_order
 						FROM $user_badge_table
-						WHERE item_id = $item_id"; // No need to check for user_id, item_id is the table's primary key.
+						WHERE item_id = $item_id";
```
Please use `WHERE item_id  = ' . (int)$item_id` to make it clear that this is safe.

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -484,9 +449,9 @@ function main($id, $mode)
 							AND item_id <> $item_id
```
Likewise here (and the line below).

pchote commented on ucp/useraccounts_module.php:
```diff
@@ -484,9 +449,9 @@ function main($id, $mode)
 							AND user_id = $this_user_id";
 						$this->db->sql_query($sql);
-						$move_executed = (bool) $this->db->sql_affectedrows();
+						$move_executed = (bool)$this->db->sql_affectedrows();
```
This cast isn't needed.

matjaeck replied:
It seems this needs to be a boolean expression for the json response to the ajax callback that triggers the reordering on the page or the js does not work.

```php
	// Response for the ajax callback that reorders items on the page.
	if ($this->request->is_ajax())
	{
		$json_response = new \phpbb\json_response;
		$json_response->send(array(
			'success' => $move_executed,
		));
	}
```

```js
phpbb.addAjaxCallback('row_down', function(res) {
	if (typeof res.success === 'undefined' || !res.success) {
		return;
	}
```

matjaeck commented:
Updated and tested, fix-up requests without reply addressed.  Since we generally want to make clear that something is safe, I included the code below (based on phpBB) again which I removed earlier.

```php
foreach ($markedkeys as $markedkey)
{
	$sql_in[] = (int)$markedkey;
}
$marked_keys_sql = ' AND ' . $this->db->sql_in_set('item_id', $sql_in);
unset($sql_in);
```

AMHOL commented on core/core.php:
```diff
@@ -118,28 +128,12 @@ public function get_update_sql($fingerprint)
 		$sql = 'UPDATE ' . $this->table_prefix . 'openra_keys' . '
 			SET ' . $this->db->sql_build_array('UPDATE', $sql_array) . '
 			WHERE fingerprint = "' . $this->db->sql_escape($fingerprint) . '"';
-		return $sql;
```
Is this intentional? Seems like a method which has a name that starts with `get_` should return something?

pchote commented on core/core.php:
```diff
@@ -41,72 +40,83 @@ public function get_info_sql($fingerprint)
 	 */
 	public function get_ubadge_sql_by_id($user_id)
 	{
+		settype($user_id, 'int');
```
Please stick with what I suggested earlier - the entire motivation for casting the variables inline is to make it clear that all input is being sanitized at the point of the sql definition, without having to carefully check how the variables are defined and used previously.

matjaeck replied:
Writing it inline didn't work for me when I tested it the first time (probably a syntax error). Then I misunderstood https://stackoverflow.com/questions/13012195/typecast-function-parameters, and erroneously concluded that I had to use the `settype()` function instead.

pchote commented on core/core.php:
```diff
@@ -153,21 +147,25 @@ public function get_ext_img_path()
 	 */
 	public function validate_badge_order($user_id)
 	{
+		settype($user_id, 'int');
```
Likewise here.

matjaeck commented:
Updated.

matjaeck commented:
Have tested  what is intended to be done in the module again and couldn't find any issues. I tried to manipulate the user input in the html or url but didn't notice any unexpected behavior. Also tested it against the current bleed and playtest in cnc and only noticed some other things that do not seem to be related to this pr.

- While the tooltip badge data updates instantly, the profile shown when in the main menu does not update.
- When only changing the URL for Profile to my local forum, it does not display badge icons
